### PR TITLE
chore: add tests, CI linting, fix lint issues

### PR DESCRIPTION
**Pull request by ankurs**
## Summary
- Add unit tests for spans, nil safety, and metadata reader/writer
- Fix unchecked `Inject()` error and empty branches
- Add golangci-lint config and CI lint job
- Add `-race` flag to test target
- Standardize CI with separate build/test/lint jobs

## Test plan
- [x] `make build` passes
- [x] `make test` passes (with -race)
- [x] `make lint` passes with 0 issues

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -22,7 +22,33 @@
...
+    - name: Set up Go
+      uses: actions/setup-go@v5
+      with:
+        go-version: "1.25"
 
     - name: Test
-      run: go test -v ./...
+      run: make test
+
+  lint:
```
[the remark this answers is not in the document]
Comment (ankurs) [reply .github/workflows/go.yml:43]: Fixed: permissions added, actions pinned to commit SHAs, go-version-file: go.mod replaces hardcoded version.

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -22,7 +22,33 @@
...
+    - name: Set up Go
+      uses: actions/setup-go@v5
+      with:
+        go-version: "1.25"
+
+    - name: Lint
+      uses: golangci/golangci-lint-action@v9
+      with:
+        version: v2.8
+        args: --timeout 5m
```
[the remark this answers is not in the document]
Comment (ankurs) [reply .github/workflows/go.yml:57]: Fixed: permissions added, actions pinned to commit SHAs, go-version-file: go.mod replaces hardcoded version.

File: .github/workflows/go.yml [hunk 3]
```diff
@@ -22,7 +22,33 @@
...
+    steps:
+    - uses: actions/checkout@v4
+
+    - name: Set up Go
+      uses: actions/setup-go@v5
+      with:
+        go-version: "1.25"
+
+    - name: Lint
+      uses: golangci/golangci-lint-action@v9
```
[the remark this answers is not in the document]
Comment (ankurs) [reply .github/workflows/go.yml]: Fixed: permissions added, actions pinned to commit SHAs, go-version-file: go.mod replaces hardcoded version.

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -22,7 +25,33 @@ jobs:
...
-      run: go build -v ./...
+      run: make build
+
+  test:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+
+    - name: Set up Go
+      uses: actions/setup-go@v5
```
[the remark this answers is not in the document]
Comment (ankurs) [reply .github/workflows/go.yml]: Fixed: permissions added, actions pinned to commit SHAs, go-version-file: go.mod replaces hardcoded version.

File: Makefile [hunk 5]
```diff
@@ -1,10 +1,16 @@
-.PHONY: build test doc
+.PHONY: build test doc lint bench
 build:
 	go build ./...
 
 test:
-	go test ./...
+	go test ./... -race
```
[the remark this answers is not in the document]
Comment (ankurs) [reply Makefile]: Fixed: -race flag moved before package pattern.

File: tracing.go [hunk 6]
```diff
@@ -296,17 +296,12 @@ func GRPCTracingSpan(operationName string, ctx context.Context) context.Context
...
+		_ = tracer.Inject(span.Context(), opentracing.TextMap, metadataReaderWriter{&md})
 	}
 
 	var span opentracing.Span
 	wireContext, err := tracer.Extract(opentracing.TextMap, metadataReaderWriter{&md})
-	if err != nil && err != opentracing.ErrSpanContextNotFound {
-		// log.Info(ctx, "err", err, "component", "tracing")
-	}
+	_ = err
 	span = tracer.StartSpan(operationName, otext.RPCServerOption(wireContext))
```
[the remark this answers is not in the document]
Comment (ankurs) [reply tracing.go]: Fixed: Extract error now handled properly — root span started on failure.

File: tracing.go [hunk 7]
```diff
@@ -235,7 +235,7 @@ func NewHTTPExternalSpan(ctx context.Context, name string, url string, hdr http.
 func traceHTTPHeaders(ctx context.Context, sp opentracing.Span, hdr http.Header) {
 	// Transmit the span's TraceContext as HTTP headers on our
 	// outbound request.
-	opentracing.GlobalTracer().Inject(
+	_ = opentracing.GlobalTracer().Inject(
 		sp.Context(),
 		opentracing.HTTPHeaders,
 		opentracing.HTTPHeadersCarrier(hdr))
```
[the remark this answers is not in the document]
Comment (ankurs) [reply tracing.go:242]: Fixed: comment added explaining best-effort inject semantics.

File: tracing_test.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	"testing"
+
+	"google.golang.org/grpc/metadata"
+)
+
+func TestNewInternalSpan(t *testing.T) {
+	ctx := context.Background()
+	span, newCtx := NewInternalSpan(ctx, "test-internal")
+	if span == nil {
+		t.Fatal("expected non-nil span")
```
[the remark this answers is not in the document]
Comment (ankurs) [reply tracing_test.go:15]: Fixed: tests pass because span constructors handle nil NR gracefully. Added TestGRPCTracingSpan to cover the Extract/StartSpan logic.

File: tracing.go [hunk 9]
```diff
@@ -296,18 +297,16 @@ func GRPCTracingSpan(operationName string, ctx context.Context) context.Context
...
 	}
 
 	var span opentracing.Span
 	wireContext, err := tracer.Extract(opentracing.TextMap, metadataReaderWriter{&md})
-	if err != nil && err != opentracing.ErrSpanContextNotFound {
-		// log.Info(ctx, "err", err, "component", "tracing")
+	if err != nil {
+		span = tracer.StartSpan(operationName)
+	} else {
+		span = tracer.StartSpan(operationName, otext.RPCServerOption(wireContext))
```
[the remark this answers is not in the document]
Comment (ankurs) [reply tracing.go:308]: Fixed: added TestGRPCTracingSpan covering both with and without parent span.

File: tracing_test.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	}
```
[the remark this answers is not in the document]
Comment (ankurs) [reply tracing_test.go:16]: Tests pass — span constructors handle nil NR app gracefully via nil checks.
